### DISCUSSION TITLE
[Feature Request] Dynamic datasource

### DISCUSSION THREAD

**Discussion by qbx2** [lowcoder-org/lowcoder#107] [Ideas]
Original issue by @devendrakhatri: https://github.com/openblocks-dev/openblocks/issues/297

For the use cases where we have multiple environments (DEV, UAT, PROD) and each environment has a dedicated database with exactly same schema.
Is it possible to build one dashboard with an option to switch between environments ?
With current datasource capability, looks like dashboard is tightly coupled with a specific database.

Can we think of passing datasource dynamically?

Probably we can call it "environment manager" which can manage all the base URLs, database,. etc. at single place and dashboard can switch to environments using dropdown list?

Environment | Service 1 | Database 1 | Database 2
-- | -- | -- | --
DEV | S1_DS_Dev | D1_DB_Dev | D2_DB_Dev
UAT | SX_DS_Uat | D1_DB_Uat | D2_DB_Uat
PROD | SY_DS_Prod | D1_DB_Prod | D2_DB_Prod

**Comment by falnyr** [comment]
FYI moving feature request into a discussion as this needs a bit more thought first.

I would not recommend Lowcoder connecting to multiple environments as this can lead to editing production data by accident. Why not having different instances of Lowcoder for each environment?

**Reply by qbx2** [reply]
Without this feature, we need to edit applications for each stack and it is not guaranteed to have same applications across the environments because it is done by humans at the moment.

**Reply by qbx2** [reply]
Anyway, we need a solution to reuse the application code for each environment.

**Reply by falnyr** [reply]
How many queries from a single data source do you have per app?

**Comment by qbx2** [comment]
10-15 queries on average.

**Reply by falnyr** [reply]
What you can do as a temporary workaround (albeit painful) is:

1. Create option element that will be your environment switcher.
2. Duplicate queries by the number of environments you have
3. Create a transformer that will run a query based on the environment selected

Your request basically describes having dynamic data sources if I understand that correctly.

**Reply by lanedsmu** [reply]
It's been awhile since there was any traffic on this particular discussion, but I'd like to chime in with some additional use cases.  
1. I've got lots of job servers, each with its own database that both controls the jobs that run on that server and which logs the results and output of each job.  I'd like to have a single app that allows the user to pick a server from a drop-down menu and then query the database on that server (using just one query) for the job(s) status(es).
2. Similar, but let's say with a web api: I've got several controllers for building automation, each with its own REST API endpoint.  I'd like to allow a user to pick the controller (building, perhaps), and then view the output from the REST query when run against that particular server
3. I'd like to give users the ability to start the same stored procedure on seveal different database instances; let's say they are three non-production servers:  dev, stage, and test. This stored procedure might be involved in change management for code deployment, or to run a set of unit tests on migrated code

If it were possible to use {{ placeholder }} values in connection strings as well as in queries, that would be pretty great.

**Reply by lanedsmu** [reply]
Another use case for dynamic data sources is connecting to the same data source, but using different database connection authentication (username/pwd) based on the user currently using LowCoder.

**Comment by FalkWolsky** [comment]
Environments will be supported in the Lowcoder EE Version.
Now... Since quite a while we mention the EE Version and still not even show up with a v0.1.alpha of it.
Reason is, that we focus much to the Community Edition to make it "best in place". 

However, we do still have development ongoing for EE version and somewhen in Summer we should finally show up with it.
Environments will be part of it.
